cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java

```diff
@@ -241,10 +263,123 @@ private static ReportGenerationRequest getReportRequest(Location reportIdDir) th
+      ReportStatus status = getReportStatus(reportIdDir);
+      Location savedFile = reportIdDir.append(SAVED_FILE);
+      // if the report is not saved, get the name of the report from the report generation request and include the
+      // expiry time in the response if the report is COMPLETED
+      if (!savedFile.exists()) {
+        Long expiryTimeSecs = null;
+        if (ReportStatus.COMPLETED.equals(status)) {
+          expiryTimeSecs = TimeUnit.MILLISECONDS.toSeconds(getExpiryTimeMillis(reportIdDir));
+        }
+        return new ReportMetaInfo(reportRequest.getName(), null, creationTime, expiryTimeSecs, status);
+      }
+      // read the report save request from the file
+      String reportSaveRequestString =
+        new String(ByteStreams.toByteArray(savedFile.getInputStream()), StandardCharsets.UTF_8);
+      ReportSaveRequest saveRequest = GSON.fromJson(reportSaveRequestString, REPORT_SAVE_REQUEST_TYPE);
+      // the report is saved, get the name and the description from the report save request
+      // and return null as expiry time
+      return new ReportMetaInfo(saveRequest.getName(), saveRequest.getDescription(), creationTime, null, status);
+    }
+
+    @DELETE
+    @Path("/reports/{report-id}")
+    public void deleteReport(HttpServiceRequest request, HttpServiceResponder responder,
+                             @PathParam("report-id") String reportId) throws IOException {
+      Location reportIdDir = getDatasetBaseLocation(ReportGenerationApp.REPORT_FILESET).append(reportId);
+      if (!reportIdDir.exists()) {
+        responder.sendError(404, String.format("Report with id %s does not exist.", reportId));
+        return;
+      }
+      ReportStatus status;
+      try {
+        status = getReportStatus(reportIdDir);
+      } catch (Exception e) {
+        LOG.error("Failed to get the status of the report {}", reportId, e);
+        responder.sendError(500, String.format("Failed to get the status of the report %s becasue of error: %s",
+                                               reportId, e.getMessage()));
+        return;
+      }
+      // if the report is already deleted, deleting the report again is not allowed
+      if (ReportStatus.DELETED.equals(status)) {
+        responder.sendError(404, String.format("Report with id %s has already been deleted", reportId));
+        return;
+      }
+      Location deletedFile = reportIdDir.append(TO_BE_DELETED_FILE);
+      if (!deletedFile.createNew()) {
+        reportIdDir.delete();
+        responder.sendError(500, "Failed to delete " + reportId);
+        return;
+      }
+      // Save the time of when this report is marked as to-be-deleted in the _TO_BE_DELETED_FILE file
+      try (PrintWriter writer = new PrintWriter(new OutputStreamWriter(deletedFile.getOutputStream(),
+                                                                       StandardCharsets.UTF_8), true)) {
+        writer.write(Long.toString(System.currentTimeMillis()));
```

Comment:
so on delete we don't delete the report but only mark it for deletion. that doesn't sound correct. Design doc mentions that it will delete the report. code needs to be fixed.

Reply:
it is marked for deletion but not immediately deleted because I want to avoid the possible conflict where the report is being read but is deleted at the same time. So I think it's better to mark it as expired and then let it to be deleted together with other expired reports after a certain period of time.

Reply:
how will you avoid the conflict then ? that conflict can happen any time you delete it. there is no point in not deleting it now. say we haven't added the expiration currently and users want to delete old reports, using this endpoint/UI to do that is much better than asking them to delete the directories manually. i think we should immediately delete the directory for this REST endpoint as mentioned in design.

Reply:
How large is a report? When reading the report, do you stream the report back or buffer in memory and then send back? If the report is relatively small and the deletion is fast, deleting in place seems make sense.

Generally we only do async deletion when deletion itself is complicated (e.g. it involves propagating the deletion to other related entities).

Reply:
reports in GB's will be considered large. currently we buffer the report in memory and send it back, but can improve it later. I'll make the deletion in place considering the conflict scenario is rare and won't actually cause much problem. This will simplify the current logic too

Reply:
Reports in GB? That's seems really huge. What's inside?

Reply:
just an extreme case like there are records with average size 400B, and 100 new records every minute, and the report is over a year. Normally the report should just be several MB's at most